packages/core-utils/src/query-gen.ts

```diff
@@ -0,0 +1,145 @@
+      .map((e1, i) => array.filter((e2, j) => i & (1 << j)))
+  );
+}
+
+export function generateCombinations(params: OTPQueryParams): OTPQueryParams[] {
+  const SIMPLIFICATIONS = {
+    AIRPLANE: "TRANSIT",
+    BICYCLE: "PERSONAL",
+    BUS: "TRANSIT",
+    CABLE_CAR: "TRANSIT",
+    CAR: "CAR",
+    FERRY: "TRANSIT",
+    FLEX: "SHARED", // TODO: this allows FLEX+WALK. Is this reasonable?
+    FUNICULAR: "TRANSIT",
+    GONDOLA: "TRANSIT",
+    RAIL: "TRANSIT",
+    SCOOTER: "PERSONAL",
+    SUBWAY: "TRANSIT",
+    TRAM: "TRANSIT",
+    TRANSIT: "TRANSIT",
+    WALK: "WALK"
+  };
+
+  const TRANSIT_SUBMODES = Object.keys(SIMPLIFICATIONS).filter(
+    mode => SIMPLIFICATIONS[mode] === "TRANSIT" && mode !== "TRANSIT"
+  );
+
+  const VALID_COMBOS = [
+    ["WALK"],
+    ["PERSONAL"],
+    ["TRANSIT", "SHARED"],
+    ["WALK", "SHARED"],
+    ["TRANSIT"],
+    ["TRANSIT", "PERSONAL"],
+    ["TRANSIT", "CAR"]
+  ];
+
+  const BANNED_TOGETHER = ["SCOOTER", "BICYCLE"];
+
+  const queryTransitSubmodes = params.modes
+    .filter(mode => TRANSIT_SUBMODES.includes(mode.mode))
+    .map(mode => mode.mode);
+
+  return (
+    combinations(params.modes)
+      .filter(combo => {
+        if (combo.length === 0) return false;
+
+        // All current qualifiers currently simplify to "SHARED"
+        const simplifiedModes = Array.from(
+          new Set(
+            combo.map(c => (c.qualifier ? "SHARED" : SIMPLIFICATIONS[c.mode]))
+          )
+        );
+
+        // Ensure that if we have one transit mode, then we include ALL transit modes
+        if (simplifiedModes.includes("TRANSIT")) {
+          const flatModes = combo.map(m => m.mode);
+          if (
+            combo.reduce((prev, cur) => {
+              if (queryTransitSubmodes.includes(cur.mode)) {
+                return prev - 1;
+              }
+              return prev;
+            }, queryTransitSubmodes.length) !== 0
+          ) {
+            return false;
+          }
+        }
+
+        // OTP doesn't support multiple non-walk modes
+        if (BANNED_TOGETHER.every(m => combo.find(c => c.mode === m)))
+          return false;
+
+        return !!VALID_COMBOS.find(
+          vc =>
+            simplifiedModes.every(m => vc.includes(m)) &&
+            vc.every(m => simplifiedModes.includes(m))
+        );
+      })
+      // create new filter that removes subTransit modes from appearing on their own
+      // ONLY IF there's multiple of them!
+      .map(combo => ({ ...params, modes: combo }))
+  );
+}
+
+// eslint-disable-next-line import/prefer-default-export
+export function generateOtp2Query(params: OTPQueryParams): any {
+  const { to, from, modeSettings } = params;
+
+  const modeSettingValues = modeSettings.reduce((prev, cur) => {
+    prev[cur.key] = cur.value;
+    return prev;
+  }, {}) as ModeSettingValues;
+
+  const {
+    walkReluctance,
+    wheelchair,
+    bikeReluctance,
+    carReluctance,
+    allowBikeRental
```

Comment:
I am wary of linking the mode setting config too rigidly to the query gen, since the mode settings could be anything and we can include custom logic here. Perhaps one way to do this would be to add a boolean property to the "default" mode setting definitions like `graphQlParam` that, if true, indicates that that mode setting's key matches exactly with the variable name that needs to be passed into the graphql query.

Reply:
yeah whatever we need to do to avoid writing this list out multiple times!

Reply:
I liked the idea of this but couldn't get importing the yml file to work in core-utils. Making this change would require moving the modeSettings defaults yml file over to core-utils. Maybe discuss offline